src/etl/impl/SubscriptionSource.cpp

```diff
@@ -0,0 +1,336 @@
+#include <openssl/err.h>
+
+#include <algorithm>
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <future>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace etl::impl {
+
+SubscriptionSource::~SubscriptionSource()
+{
+    stop();
+
+    if (wsConnection_) {
+        boost::asio::spawn(strand_, [this](boost::asio::yield_context yield) { wsConnection_->close(yield); });
+    }
+    // To be sure that all the async operations are completed
+    std::future<void> future = boost::asio::post(strand_, boost::asio::use_future);
+    future.get();
+
+    wsConnection_.reset();
+}
+
+void
+SubscriptionSource::run()
+{
+    subscribe();
+}
+
+bool
+SubscriptionSource::hasLedger(uint32_t sequence) const
+{
+    auto validatedLedgersData = validatedLedgersData_.lock();
+    for (auto& pair : validatedLedgersData->validatedLedgers) {
+        if (sequence >= pair.first && sequence <= pair.second) {
+            return true;
+        }
+        if (sequence < pair.first) {
+            // validatedLedgers_ is a sorted list of disjoint ranges
+            // if the sequence comes before this range, the sequence will
+            // come before all subsequent ranges
+            return false;
+        }
+    }
+    return false;
+}
+
+bool
+SubscriptionSource::isConnected() const
+{
+    return isConnected_;
+}
+
+void
+SubscriptionSource::setForwarding(bool isForwarding)
+{
+    isForwarding_ = isForwarding;
+}
+
+std::chrono::steady_clock::time_point
+SubscriptionSource::lastMessageTime() const
+{
+    return lastMessageTime_.lock().get();
+}
+
+std::string const&
+SubscriptionSource::validatedRange() const
+{
+    return validatedLedgersData_.lock()->validatedLedgersRaw;
+}
+
+void
+SubscriptionSource::stop()
+{
+    stop_ = true;
+}
+
+void
+SubscriptionSource::subscribe()
+{
+    boost::asio::spawn(strand_, [this, _ = boost::asio::make_work_guard(strand_)](boost::asio::yield_context yield) {
+        auto connection = wsConnectionBuilder_.connect(yield);
+        if (not connection) {
+            handleError(connection.error(), yield);
+            return;
+        }
+
+        wsConnection_ = std::move(connection).value();
+        isConnected_ = true;
+
+        auto const& subscribeCommand = getSubscribeCommandJson();
+        auto const writeErrorOpt = wsConnection_->write(subscribeCommand, yield);
+        if (writeErrorOpt) {
+            handleError(writeErrorOpt.value(), yield);
+            return;
+        }
+
+        retry_.reset();
+
+        while (!stop_) {
+            auto message = wsConnection_->read(yield);
+            if (not message) {
+                handleError(message.error(), yield);
+                return;
+            }
+
+            auto handleErrorOpt = handleMessage(message.value());
+            if (handleErrorOpt) {
+                handleError(handleErrorOpt.value(), yield);
```

Comment:
I am thinking if handleMessage raises error , do we need to retry connect? How does retry help?

Reply:
I was thinking if there is an error on any stage we should start from scratch. There is a hope that if we started to receive wrong messages it could be better with a fresh connection. I think the same logic was in the old code.